OutOfSchool/OutOfSchool.BusinessLogic/Services/Memento/MementoService.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using OutOfSchool.BusinessLogic.Services.Memento.Interfaces;
+
+namespace OutOfSchool.BusinessLogic.Services.Memento;
+
+/// <summary>
+/// Implements the IMementoService{T} interface with CRUD functionality to store an entity of type T in a cache.
+/// </summary>
+/// <typeparam name="T">T is the entity type that should be stored in the cache.</typeparam>
+public class MementoService<T> : IMementoService<T>
+{
+    private readonly ICrudCacheService crudCacheService;
+    private readonly ILogger<MementoService<T>> logger;
+
+    /// <summary>Initializes a new instance of the <see cref="MementoService{T}" /> class.</summary>
+    /// <param name="crudCacheService">The CRUD cache service.</param>
+    /// <param name="logger">The logger.</param>
+    public MementoService(
+        ICrudCacheService crudCacheService,
+        ILogger<MementoService<T>> logger)
+    {
+        this.crudCacheService = crudCacheService;
+        this.logger = logger;
+    }
+
+    /// <summary>Restores the memento.</summary>
+    /// <param name="key">The key.</param>
+    /// <returns> Representing the asynchronous operation with result of T type.</returns>
+    public async Task<T> RestoreAsync([NotNull] string key)
+    {
+        var mementoKey = key is not null ? GetMementoKey(key) : throw new ArgumentNullException(nameof(key));
```

Comment:
> 1. please move arguments validation logic out from the business logic.
> 2. Please use `ArgumentException.ThrowIfNullOrEmpty(nameof(key))` instead.
> 
> It will bring more readability to the code. Update all other places.

1) When we discussed last time this code I understood that I had to check all arguments that input to service methods.

2) Ok, I changed code